packages/storage-plus/src/keys.rs

```diff
@@ -51,9 +61,13 @@ impl<'a> PrimaryKey<'a> for &'a str {
 }
 
 // use generics for combining there - so we can use &[u8], Vec<u8>, or IntKey
-impl<'a, T: PrimaryKey<'a> + Prefixer<'a>, U: PrimaryKey<'a>> PrimaryKey<'a> for (T, U) {
+impl<'a, T: PrimaryKey<'a> + Prefixer<'a> + KeyDeserialize, U: PrimaryKey<'a> + KeyDeserialize>
+    PrimaryKey<'a> for (T, U)
+{
     type Prefix = T;
     type SubPrefix = ();
+    type Suffix = U;
+    type SuperSuffix = Self;
```

Comment:
Could you explain to me how does this works?
This is a `impl PrimaryKey` for tuple `(T, U)`, so for `(Prefix, Suffix)` - but `SuperSuffix` type is `Self`, which means `(Prefix, Suffix)`?
So the `SuperSuffix` is actually whole key?

Reply:
Yes, in this case `SuperSuffix` is the whole key. To summarize, for `(T, U)`:

- `T`: Prefix. Prefix is eager. That is, it's always "one less" than the full key.
- `U`: Suffix: Suffix is the complement of prefix.
- `()`: Sub-prefix. "One less" than Prefix. This is tantamount of calling `range` over the full key in this case.
- `(T, U)`: Super-suffix: One more than suffix. Super-suffix is the complement of sub-prefix.

Reply:
Good explanation - I would appreciate having it as a doc-comment.

Reply:
Will do.

Reply:
To be honest, I can follow this logic (`prefix` is "one-less", suffix is that "one-less"), but it seems rather arbitrary than intuitive for me.
When I read that first time I though that in case `ABCD` - `A` would be a prefix, `D` would be a suffix (instead of `ABC` as prefix).

Reply:
This is for historical reasons. We can change it (`Prefix` becomes `SubPrefix`; then we'll have `SuperPrefix` and `SubSuffix` ). But still, it would be equaly arbitrary.

Anyway, this is mostly internal stuff. The only things you see as a user are the `prefix()` and `sub_prefix()` methods.

We don't store anything except for the full key, but build the partial keys using these types as references.

Reply:
Plus, we need the `*Suffix` stuff for annotating key deserialization (sub) types.

Reply:
https://github.com/CosmWasm/cw-plus/blob/af447a055977605fd21d5893a2194e93d1b0760e/packages/storage-plus/src/map.rs#L115-L128

See how a sub-prefix is returning a super-suffix, and a prefix is returning a suffix? That means: I fixed this part, give me the rest (deserialized).

Reply:
I like the (new) doc comments above and would consider this resolved.
Maybe those trait types could use more docstrings?

```rust
type Suffix: KeyDeserialize;
type SuperSuffix: KeyDeserialize;
```

They make sense to me, but I am also guilty for inventing half the names.